### ISSUE TITLE
Ignore everything but a given selector

### ISSUE THREAD

**Issue by GlynnPhillips** [springernature/pa11y#133]
Previously raised by @davidjb in https://github.com/springernature/pa11y/issues/108

> +1. In my case I'm interested in also ignoring everything but a specific selector, so perhaps include/exclude options would satisfy both. Having pa11y test the whole page but just filter the report output could be a solution.

> In my specific use case, I have documentation pages from an upstream web framework and want to test a specific selector on each page (eg excluding everything else).

One way of achieving this could be to altering the second option [here](https://github.com/nature/pa11y/blob/master/lib/inject.js#L16) to use a query selector instead of the whole document

**Comment by GlynnPhillips** [comment]
@nickcall and I have looked at this today and found a slightly problem with the proposed approach.

By altering the second option [here](https://github.com/nature/pa11y/blob/master/lib/inject.js#L16) to select the area of the page to test you could end up with false errors in certain examples.

One example would be using anchors to jump to another region on the page. If the anchor as within the testing area but the target element is outside of the testing area Pa11y will still return an area that the named anchor does not exist.

The alternative approach would be to filter the results after the full page has been tested however this would be considerably more work and the filtering would be an expensive process.

@rowanmanning what do you think?

**Comment by GlynnPhillips** [comment]
One approach we have come up with to achieve this without too much complexity would require pa11y to dictate that the element we want to limit our results too must have a data attribute applied to it. This data attribute could be named anything the user wants and would be passed in as an option. 

Quick example of how `/lib/inject.js` could look

Inside the [buildElementIdentifier function](https://github.com/springernature/pa11y/blob/master/lib/inject.js#L72)

```js
function buildElementIdentifier(element) {
    if(element.getAttribute(options.testArea)) {
         return options.testArea;
    }
    // The rest of this function would stay the same
}
```

So if `options.testArea` equalled `data-test-region` the `message.selector` would look something like this `html > body > data-test-region`

The [isMessageWanted functuion](https://github.com/springernature/pa11y/blob/master/lib/inject.js#L103) could then easily filter out unwanted messages without at any point in this process needing to keep all the properties of an element to filter on.

A couple of gut feelings I have about this approach is that it might not be overly elegant and also we would be enforcing the use of data attributes (or a similar attribute)

**Comment by rowanmanning** [comment]
Alternatively you could allow any selector like we discussed before but still run the test on `document`. Then after the test has run you'd have to loop over the results and exclude any relating to elements that aren't children of the specified element. Does that make sense?

Some example pseudo-code to try and help explain:

```js
function processMessages(messages) {
	return messages.filter(isMessageInTestArea).map(processMessage).filter(isMessageWanted);
}

function isMessageInTestArea(message) {
    var testAreaElement = document.querySelector(options.testArea);
    return isElementInsideAnotherElementOrWhatever(message.element, testAreaElement);
}
```

I don't like the idea of you having to make modifications to your page to support a pa11y feature.

**Comment by GlynnPhillips** [comment]
Yeah, I agree and thats what I didn't like about my solution. 

#138 now adds similar to the above but I would like some feedback on the `selector` option. Not sure if anyone has a better name for this

**Comment by GlynnPhillips** [comment]
This has now been released as part of `3.6.0` https://github.com/springernature/pa11y#rootelement-element
